Create darjutak-2018-november-sip-minutes.md (#1226)
scala/docs.scala-lang: 2018-11-26-sip-minutes.md

---
layout: sips
title: SIP Meeting Minutes - November 1-3 2018

partof: minutes
---

# Minutes

> The following agenda was distributed to attendees, please [click here](https://docs.google.com/spreadsheets/d/1JdTyAxqqKqXtiXnZBPTaesMrWk7lyTWna1VPL3N5KWo/edit?usp=sharing).

Reviewer: Could this just be

```suggestion
The agenda was distributed to attendees is [here](https://docs.google.com/spreadsheets/d/1JdTyAxqqKqXtiXnZBPTaesMrWk7lyTWna1VPL3N5KWo/edit?usp=sharing).
```

?

Author: Thanks for the wonderful feedback @hrhino !

* Jorge Vicente Cantero ([@jvican](https://github.com/jvican)), Process Lead
* Seth Tisue ([@SethTisue](https://github.com/SethTisue)), Lightbend
* Iulian Dragos ([@dragos](https://github.com/dragos)), Triplequote
* Josh Suereth ([@jsuereth](https://github.com/jsuereth)), Independent
* Sébastien Doeraene ([@sjrd](https://github.com/sjrd)), Scala Center
* Adriaan Moors ([@adriaanm](https://github.com/adriaanm)), Lightbend
* Miles Sabin ([@milessabin](https://github.com/milessabin)), Independent
* Darja Jovanovic ([@darjutak](https://github.com/darjutak)), Scala Center

## Joined online

* Eugene Burmako ([@xeno-by](https://github.com/xeno-by)), Twitter

## Not present

* Heather Miller ([@heathermiller](https://github.com/heathermiller)), CMU

## Proceedings
### Opening Remarks

The SIP Committee gathered for the first time face-to-face, for an extensive 3-day SIP meeting. The main goals and achievements of these meetings were:

- Understand better the upcoming changes to Scala; 
- Agree about the role of the Committee in the Scala 2 to Scala 3 transition;
- Outline an action plan within a set time-frame;
- Other: Unanimously voted for Guillaume Martres to join the Committee.

> *Better understanding* was enabled by in depth presentations and Q&As with the EPFL Dotty team; *the Approach* was agreed upon the first day which resulted in creating “FAQs about Scala 3” (see below) and *the Action Plan* was outlined and is still under construction; sveveral issues were opened and a project plan Meta programming in Scala 3 has been developed (please see the list at the end of this document).

Reviewer: `s/sveveral/several/`

```suggestion
*Better understanding* was enabled by in depth presentations and Q&As with the EPFL Dotty team; *the Approach* was agreed upon the first day which resulted in creating “FAQs about Scala 3” (see below) and *the Action Plan* was outlined and is still under construction; several issues were opened and a project plan "Meta-programming in Scala 3" has been developed (please see the list at the end of this document).
```

> *Better understanding* was enabled by in depth presentations and Q&As with the EPFL Dotty team; *the Approach* was agreed upon the first day which resulted in creating “FAQs about Scala 3” (see below) and *the Action Plan* was outlined and is still under construction; sveveral issues were opened and a project plan Meta programming in Scala 3 has been developed (please see the list at the end of this document).

Reviewer: quote `"Meta Programming in Scala 3"`? also, I don't see anywhere that this document is linked.

As the most important points and summary is reflected in “FAQs about Scala 3”, it will stand as an official “minutes” for this unique SIP meeting.

> The following document is not intended to fully answer the listed questions, rather to acknowledge them; the SIP Committee is outlining a frame and will pursue answers moving forward. Answers may change down the line, as the Process evolves. 

Reviewer: ```suggestion
The following document is not intended to fully answer the listed questions, but to acknowledge them; the SIP Committee is outlining a frame and will pursue answers moving forward. Answers may change down the line, as the Process evolves. 
```

"Rather" usually doesn't introduce a subordinate clause.

The second phase will continue into 2020, where the focus will shift to hardening, tooling and polishing documentation to ensure a smooth migration. 

### Who’s behind Scala 3?
Over the last five years, Prof. Martin Odersky and his team at EPFL have developed Scala 3. (Also, several Scala 2 features were first incubated in Scala 3.)
Scala 3 will continue to leverage this collaboration with the Scala 2 team at Lightbend to improve the language for existing Scala users and ease migration. 

### What’s the role of the SIP Committee in the Scala 3 migration?
The SIP Committee takes account of the interests of the varied stakeholders of the Scala language. Regarding the transition to Scala 3, the Committee is tasked with:
1. Curating language changes in the Scala specification;
2. Providing recommendations for migration from Scala 2 to Scala 3;
3. Work with the Scala Center to be a point of reference during the transition to Scala 3, incorporating feedback from implementation experience and from the community.

> During the meeting, the Committee came up with the curated list of the Core Scala 3 features, that are considered as a bundle that represents Scala 3, others are marked as "not core" or "essential". Please see the [full list here](https://docs.google.com/spreadsheets/d/1GWJUo0U3JbBtrfg5vqgb6H5S6wlU5HnTxebLcHwD1zw/edit?usp=sharing), naming the "core" features as follows:

Reviewer: You can split this into two sentences.

```suggestion
During the meeting, the Committee came up with the curated list of the Core Scala 3 features, that are considered as a bundle that represents Scala 3. Other features are marked as "not core" or "essential". Please see the [full list here](https://docs.google.com/spreadsheets/d/1GWJUo0U3JbBtrfg5vqgb6H5S6wlU5HnTxebLcHwD1zw/edit?usp=sharing), naming the "core" features as follows:
```

What is the difference between "core" and "essential"?

Reviewer 2: "Core" are well-defined features or changes that are already designed and implemented. We have now accepted them on principle, but the finer details are still up for discussion.

"Essential" are wide, *unsolved* areas for which the committee recognizes that we cannot decently ship Scala 3.0 without a final design and implementation to solve them. Nothing has been accepted yet because they're not even fully *designed* yet.

Reviewer: Thanks. That might be useful information to include. (Unless it was somewhere else and I missed it...)

Author: Agreed, I'll add that part.
We had a public SIP meeting yesterday and we spoke about it more, so this part will be shown in the upcoming minutes too.
Thanks, @sjrd for the answer and (again) @hrhino for taking the time to go as detailed over this, I truly appreciate it!!!

[Type Inference](https://dotty.epfl.ch/docs/reference/changed/type-inference.html)

[Implicit Resolution](https://dotty.epfl.ch/docs/reference/changed/implicit-resolution.html)

[Pattern matching](https://dotty.epfl.ch/docs/reference/changed/pattern-matching.html)

[Existential Types](https://dotty.epfl.ch/docs/reference/dropped/existential-types.html)

[Type Projection](https://dotty.epfl.ch/docs/reference/dropped/type-projection.html)

[Class Shadowing](https://dotty.epfl.ch/docs/reference/dropped/class-shadowing.html)

[Trait Parameters](https://dotty.epfl.ch/docs/reference/trait-parameters.html)

### What’s the plan for keeping the migration period as short as possible?
A smooth migration process is key for the success of Scala 3. We have learned from our own past experience as well as that of other language communities (for example, Python 3) to recommend the following plan.

These are the key properties of a successful upgrade plan that we recommend to incorporate in the migration to Scala 3.

1. **Incremental.** Instead of a one-time big upgrade, users should be able to adopt Scala 3 at their own pace.
  

>   a) Compatibility and cross-building. Users should be able to use a common subset of Scala to mix Scala 2 and Scala 3 projects (tbd after the feature freeze) in the same codebase. Scala 2 should guide users towards this subset through deprecation warnings.

Reviewer: ```suggestion
  a) Compatibility and cross-building. Users should be able to use a common subset of Scala to mix Scala 2 and Scala 3 projects (to be determined after the feature freeze) in the same codebase. Scala 2 should guide users towards this subset through deprecation warnings.
```

either capitalize the acronym or write it out.

  
  b) [Tasty](https://www.scala-lang.org/blog/2018/04/30/in-a-nutshell.html) compatibility. Scala 2 and Scala 3 should converge on the use of Tasty, an intermediate representation format, that will have a strong backwards compatibility policy.

> 2. **Static types.** Most Scala code leverages static types which makes migrating safer and easier.

Reviewer: Is this a "propert[y] of a successful upgrade plan"? It seems more of a facet of the language that could help, but any upgrade plan would be focused on a language with static types.

Reviewer 2: That's a good point. Strictly speaking it shouldn't be here if we only talk of the upgrade plan itself. However, we (the committee) believe that it is essential to point out static types as a major facilitating factor in the migration, especially wrt. the unavoidable comparison with Python 3.

Perhaps we should move this point in the "In addition" section below the bullets?

3. **Tooling to ease upgrades.** There should be a rewriting tool for automatic migrations and changes in the current Scala tooling. This tool should accommodate most of the needed changes.

> 4. **Shared standard library.** Scala 2.14 and Scala 3.0 should share the same Scala standard library.

Reviewer: doesn't Dotty compile with the library from 2.13? Are there 2.13 library changes that aren't in Dotty or _vice versa_?

Reviewer 2: There are currently some Dotty-specific classes in its library, I believe. But for the most part, it currently transitively depend on Scala 2.12's library. Nevertheless, this is an important property that we should not break, even when Dotty *fully* bootstraps.

> In addition, the Scala community has a culture of upgrading the ecosystem on every major version of Scala through the community build. We  recommend to leverage the same mechanism to evaluate Scala upgrades with the migration to Scala 3.

Reviewer: ```suggestion
In addition, the Scala community has a culture of upgrading the ecosystem on every major version of Scala through the community build. We intend to leverage the same mechanism to evaluate Scala upgrades with the migration to Scala 3.
```

Also, the word "evaluate" sounds strange to me here. Is this saying that the migration will be something that Scala users are used to?

Reviewer 2: "intend" is nice.
I also suggest "evaluate" -> "vet". Does that answer your question @hrhino?

### How will the transition to Scala 3 affect users of Scala 2 macros and reflection?

> The dependency of Scala 2 macros and reflection on internal implementation details of the Scala 2 compiler mean that significant change is inevitable if Scala is to evolve.

Reviewer: ```suggestion
The dependency of Scala 2 macros and reflection on internal implementation details of the Scala 2 compiler means that significant change is inevitable if Scala is to evolve.
```

"dependency" is singular.

> We recognize that important parts of the Scala ecosystem have made essential use of the Scala 2 facilities and that it is vital that as many as possible of these use cases be accommodated in Scala 3 in some form or another. This will be disruptive, but we hope to mitigate the disruption by providing facilities which make the more straightforward and important scenarios simpler whilst leaving others still possible.

Reviewer: "whilst" is a bit archaic. could just be "while"

```suggestion
We recognize that important parts of the Scala ecosystem have made essential use of the Scala 2 facilities and that it is vital that as many as possible of these use cases be accommodated in Scala 3 in some form or another. This will be disruptive, but we hope to mitigate the disruption by providing facilities which make the more straightforward and important scenarios simpler while still leaving others possible.
```

> Our direction is still evolving, however we believe that replacing the current excessively general and expressive macro system with a suite of less powerful but complementary tools is the way forward.

Reviewer: ```suggestion
Our direction is still evolving; however we believe that replacing the current excessively general and expressive macro system with a suite of less powerful but complementary tools is the way forward.
```

"however" is an adverb, not a conjunction.

Also, "complementary" to what? Each other?

> Currently we are exploring options which range from improved support for type level programming in the language itself (eg. specialized inline, match types, stable definitions, GADT improvements); intensifying certain features currently supported by macros (eg. by-name implicits, generic programming primitives); through to less general forms of metaprogramming (quote/splice and staging) and portable reflection via TASTy (which we recommend to support in both Scala 2/3 and via compiler-independent libraries and tools; [see also](https://github.com/scalacenter/advisoryboard/pull/40)). We recommend that most current uses of Scala macros and reflection can be accommodated by some combination of these tools.

Reviewer: The capitalization of "TASTy" is inconsistent throughout.

Thould "intensifying" be "intrinsifying"?

The "see also" could just be a link to the recommendation on the text "we recommend" or similar.

> For more about the project progress, please see https://github.com/lampepfl/dotty/issues/5489

Reviewer: ```suggestion
For more about the project's progress, please see https://github.com/lampepfl/dotty/issues/5489
```
